src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/SemanticTokens.cs

```diff
@@ -77,4 +81,135 @@ internal partial class RazorCustomMessageTarget
+        var lspMethodName = Methods.TextDocumentSemanticTokensRangeName;
+
+        foreach (var range in semanticTokensParams.Ranges)
+        {
+            var newParams = new SemanticTokensRangeParams
+            {
+                TextDocument = semanticTokensParams.TextDocument,
+                PartialResultToken = semanticTokensParams.PartialResultToken,
+                Range = range,
+            };
+
+            using var _ = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId);
+            var task = _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensRangeParams, VSSemanticTokensResponse>(
+                textBuffer,
+                lspMethodName,
+                languageServerName,
+                newParams,
+                cancellationToken);
+            requestTasks.Add(task);
+        }
+
+        var results = await Task.WhenAll(requestTasks).ConfigureAwait(false);
+        var nonEmptyResults = results.Select(r => r?.Response?.Data).WithoutNull().ToArray();
+
+        if (nonEmptyResults.Length != semanticTokensParams.Ranges.Length)
+        {
+            // Weren't able to re-invoke C# semantic tokens but we have to indicate it's due to out of sync by providing the old version
+            return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: csharpDoc.HostDocumentSyncVersion);
+        }
+
+        var data = StitchSemanticTokenResponsesTogether(nonEmptyResults);
+
+        var response = new ProvideSemanticTokensResponse(data, semanticTokensParams.RequiredHostDocumentVersion);
+
+        return response;
+    }
+
+    private int[] StitchSemanticTokenResponsesTogether(int[][] responseData)
+    {
+        var count = responseData.Sum(r => r.Length);
+        var data = new int[count];
+        var dataIndex = 0;
+        var lastTokenLine = 0;
+
+        for (var i = 0; i < responseData.Length; i++)
+        {
+            var curData = responseData[i];
+
+            if (curData.Length == 0)
+            {
+                continue;
+            }
+
+            Array.Copy(curData, 0, data, dataIndex, curData.Length);
+            if (i != 0)
+            {
+                // The first two items in result.Data will potentially need it's line/col offset modified
+                var lineDelta = data[dataIndex] - lastTokenLine;
+
+                // Update the first line copied over from curData
+                data[dataIndex] = lineDelta;
+
+                // Update the first column copied over from curData if on the same line as the previous token
+                if (lineDelta == 0)
+                {
+                    var lastTokenCol = 0;
+
+                    // Walk back accumulating column deltas until we find a start column (indicated by it's line offset being non-zero)
+                    for (var j = dataIndex - RazorSemanticTokensInfoService.TokenSize; j >= 0; j -= RazorSemanticTokensInfoService.TokenSize)
```

Comment:
Nit: This whole algorithm relies on knowing that the data is in 5-int form, and would break spectacularly if the format were to change, so sharing the constant `5` seems a little pathological 😛

Reply:
The algm depends on the fact that the line/col entries are at position 0/1 of each item's data, not really on the fact that each item has 5 entries. But yes, probably a little pathological.